grr/server/grr_response_server/databases/spanner_utils.py

```diff
@@ -0,0 +1,477 @@
+    query itself have to be escaped.
+
+    Also, the query literal is not allowed to contain any '@p{idx}' strings
+    inside as that would lead to an incorrect behaviour when evaluating the
+    query. To prevent mistakes the function will raise an exception in such
+    cases.
+
+    [1]: https://docs.python.org/3/library/stdtypes.html#str.format
+
+    Args:
+      query: An SQL string with parameter placeholders.
+      params: A dictionary mapping parameter name to a value.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A cursor over the query results.
+
+    Raises:
+      ValueError: If the query contains disallowed sequences.
+      KeyError: If some parameter is not specified.
+    """
+     if not param_type:
+      param_type = {}
+    names, values = collection.Unzip(params.items())
+    query =  self._parametrize(query, names)
+
+    for key, value in params.items():
+      if key not in param_type:
+        try:
+          param_type[key] = self._get_param_type(value)
+        except TypeError as e:
+          print(f"Warning for key '{key}': {e}. Setting type to None.")
+          param_type[key] = None # Or re-raise, or handle differently
+
+    with self._pyspanner.snapshot() as snapshot:
+        results = snapshot.execute_sql(
+            query,
+            params=params,
+            param_types=param_type,
+            request_options={"request_tag": txn_tag}
+        )
+
+    return results
+
+  def ParamQuerySingle(
+      self, query: str, params: Mapping[str, Any],
+      param_type: Optional[dict] = None, txn_tag: Optional[str] = None
+  ) -> Row:
+    """Queries the database for a single row using with a query with params.
+
+    See documentation for `ParamQuery` to learn more about the syntax of query
+    parameters and other caveats.
+
+    Args:
+      query: An SQL string with parameter placeholders.
+      params: A dictionary mapping parameter name to a value.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A single result of running the query.
+
+    Raises:
+      NotFound: If the query did not return any results.
+      ValueError: If the query yielded more than one result.
+      ValueError: If the query contains disallowed sequences.
+      KeyError: If some parameter is not specified.
+    """
+    return self.ParamQuery(query, params, param_type=param_type, txn_tag=txn_tag).one()
+
+  def ParamExecute(
+      self, query: str, params: Mapping[str, Any], txn_tag: Optional[str] = None
+  ) -> None:
+    """Executes the given query with parameters against a Spanner database.
+
+    Args:
+      query: An SQL string with parameter placeholders.
+      params: A dictionary mapping parameter name to a value.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      Nothing.
+
+    Raises:
+      ValueError: If the query contains disallowed sequences.
+      KeyError: If some parameter is not specified.
+    """
+    names, values = collection.Unzip(params.items())
```

Comment:
```suggestion
    names, _ = collection.Unzip(params.items())
```

`values` isn't being used. Let's omit it.